### PR TITLE
[flang][debug] Support fixed size character type.

### PR DESCRIPTION
**Pull request by abidh**
This PR adds debug support for fixed size character type. The character type gets translated to DIStringType.

As I have noticed in comments, currently DIStringType does not have a way to represent the underlying character type of the string. This restricts our ability to represent wide string. As an example, this is how the debugger shows 2 different type of string. Note that non-ascii characters work ok with default kind string.

  character(kind=4, len=5) :: str1
  character(len=16) :: str2
  str1 = 'hello'
  str2 = 'π = 3.14'

(gdb) p str1
$1 = 'h\000\000\000e\000\000\000l\000\000\000l\000\000\000o\000\000\000'

(gdb) p str2
$2 = 'π = 3.14       '

### CODE REVIEW

File: flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp [hunk 1]
```diff
@@ -190,6 +190,33 @@ mlir::LLVM::DITypeAttr DebugTypeGenerator::convertSequenceType(
...
+  if (charTy.getFKind() != 1)
+    encoding = llvm::dwarf::DW_ATE_UCS;
+
+  // FIXME: Currently the DIStringType in llvm does not have the option to set
+  // type of the underlying character. This restricts out ability to represent
+  // string with non-default characters. Please see issue #95440 for more
+  // details.
+  return mlir::LLVM::DIStringTypeAttr::get(
+      context, llvm::dwarf::DW_TAG_string_type,
+      mlir::StringAttr::get(context, ""),
```
Comment (tblah) [inline flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:214]: What is this empty string used for?

Comment (abidh) [reply flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:214]: It is name of the type. We currently don't give it a name so debugger makes one itself.

```
character(len=20) :: second

(GDB) ptype second
type = character*20
```
If we want, we could pass a different name like "character (kind=1, len=20)" and then debugger will show that instead. 
For most of the types, we are not passing a name at the moment as debugger can mostly generate name from the information passed. For things like derived types, we will have to pass the name.

Comment (tblah) [reply flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:214]: Thanks!

**Review by tblah — APPROVED** [review]
LGTM, thanks!

File: flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp [hunk 2]
```diff
@@ -190,6 +190,33 @@ mlir::LLVM::DITypeAttr DebugTypeGenerator::convertSequenceType(
...
+      charTy.getLen() * kindMapping.getCharacterBitsize(charTy.getFKind()),
+      /* alignInBits */ 0, /* stringLength */ nullptr,
```
Comment (vzakhari) [inline flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:216]: nit: the usual style for these comments is `/*alignBits=*/` (no spaces and `=` at the end).

**Review by vzakhari — APPROVED** [review]
LGTM. Thank you!

**Review by jeanPerier — APPROVED** [review]
LGTM

File: flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp (hunk 2 above)
Comment (abidh) [reply flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:216]: I will fix it here and throughout the file in a separete PR. Thanks.

**Pull request merged**